### PR TITLE
[18][8][9] Version 1 functionality implemented

### PR DESCRIPTION
**Pull request by AZielinsky95**
### Summary

This PR implements the functionality of getting the history of commits if the Coda table is empty and a fromTag is provided. 
If not fromTag is provided and the table is empty it will simply log the one commit event. For any subsequent commits it will grab the latest timestamp from the table and backfill all commits since but not including the last commit written to the table. 

All inputs required for the workflow are added to action.yml 

Further steps can be taken to clean up the code. Like turning commits api into a class so that we don't have to create a new octokit instance for each method. Since this is working I'm just submitting the PR in this state. 

If you want to test this action 
- Add the workflow posted below to your repo
- Add a secret to your repo for your Coda API token called CODA_TOKEN 
- Add a secret to your repo for your  Personal Access Token  called PAT
- Get your Coda doc Id. [Link to Coda doc id extractor](https://coda.io/developers/apis/v1#section/Using-the-API/Resource-IDs-and-Links)
- Set the fromTag to be either  your last release tag or remove the property 
- There's also a table property that you can add to specify a table name to write too ( default is "Unreleased Changes")
- Commit the workflow and it will write to your Coda doc!

````
name: coda-changelog-action

# Controls when the workflow will run
on:
  # Triggers the workflow on push or pull request events but only for the main branch
  push:
    branches: [ main ]

  # Allows you to run this workflow manually from the Actions tab
  workflow_dispatch:

jobs:
  upload-to-coda:
    # The type of runner that the job will run on
    runs-on: ubuntu-latest

    # Steps represent a sequence of tasks that will be executed as part of the job
    steps:
      - name: Extract branch name
        run: echo "::set-output name=branch::$(echo ${GITHUB_REF#refs/heads/})"
        shell: bash
        id: extract_branch
        
      - name: Update Changelog  
        id: update-changelog
        uses: steamclock/coda-changelog@v1
        with:
          fromTag: 'v3.0'
          commits: ${{ toJSON(github.event.commits) }}
          token: ${{ secrets.PAT }}
          coda-token: ${{ secrets.CODA_TOKEN }}
          doc-id: 'p-_LNphf30'
          branch: ${{ steps.extract_branch.outputs.branch }}
        env:
         GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}

````
Screen Shot 2021-10-20 at 10 54 52 AM

Side note: Prior to finding the needed endpoint in the github API I spent a little too long getting a shell command to output the history of logs since a tag and format it into the correct JSON format to be consumed by the action. Just wanted to share cause I thought it was cool haha. 

`  echo "::set-output name=commits::$(echo $(git log ${{ <tag name> }}.. \
              --pretty=format:'{%n  "commit": "%H",%n "body": "%b",%n  "message": "%s",%n    "timestamp": "%aD"%n,%n  "author": {%n    "name": "%aN",%n    "email": "%aE" } %n},' \
              $@  | awk 'BEGIN { print("[") } { print($0) } END { print("]") }' | python -u -c \
              'import ast,json,sys; print(json.dumps(ast.literal_eval(sys.stdin.read())))'))"`

### CODE REVIEW

File: src/main.ts [hunk 1]
```diff
@@ -1,19 +1,65 @@
...
+      commitsToUpload = commitEvent
+    } else {
+      //If we have a lastCommit date value get all commits since the date 
+      console.log(`Fetching Commit History since date: ${lastCommitDate}`)
+      commitsToUpload = await commits.getCommitsSinceDate(token,owner,repo, lastCommitDate)
+    }
+    console.log(`# of commits found: ${commitsToUpload.length}`)
+    core.endGroup()
+
+    var columns = await api.getColumnsForTable(docId, tableName)
```
Comment (jenncoop) [inline src/main.ts:54]: Should this line be moved into the else block? Or is it being used elsewhere?

File: src/api/commits.ts [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+                    author: {
+                        email: c.commit.author?.email ,
+                        name: c.commit.author?.name,
+                        username: c.author?.login
+                    },
+                    message: c.commit.message,
+                    timestamp: c.commit.author?.date,
+                    url: c.commit.url
+                } as Commit
+            })
```
Comment (jenncoop) [inline src/api/commits.ts:64]: Would it be worth DRYing this up a bit and moving the author generation into its own function?

**Review by jenncoop — APPROVED** [review]
I haven't had a chance to try it out yet, but so far it looks really good! I added a number of questions/suggestions related to JS mostly, but given how early this project is these are non blocking.

File: src/api/coda.ts [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        .catch((error: any) => {
+            core.warning(error);
+        });
+}
+
+export async function insertRows(docId: string, tableName: string, rows: Rows) {
+    return axios 
+        .post(`docs/${docId}/tables/${tableName}/rows`, rows)
+        .then(async (response: any) => {
+            console.log(response)
```
[the remark this answers is not in the document]
Comment (AZielinsky95) [reply src/api/coda.ts:24]: It was! I'll remove it.

File: src/row-builder.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    commits.forEach((commit) => {
+        var cells: TableCell[] = []
+        for(var key in columns) {
+            var column = columns[key]
+            var cell = {
+                column: column.id,
+                value: valueForColumn(column.name, commit)
+            } as TableCell
+            cells.push(cell)
+        }
```
[the remark this answers is not in the document]
Comment (AZielinsky95) [reply src/row-builder.ts:17]: Nice!

**Pull request merged**